### DISCUSSION TITLE
Indexing and slicing with brackets

### DISCUSSION THREAD

**Discussion by vanillajonathan** [PRQL/prql#1811] [Ideas]
Python lets you do string slicing using an indexer with bracket syntax like this:

```pycon
In [1]: hamburger = "hamburger"

In [2]: hamburger[3]
Out[2]: 'b'

In [3]: hamburger[3:]
Out[3]: 'burger'

In [4]: hamburger[3:6]
Out[4]: 'bur'

In [5]: hamburger[:6]
Out[5]: 'hambur'
```

SQL provides the `SUBSTRING()` function so PRQL could impement a simliar indexer using brackets as in Python and translate that into SQL.

Docs: https://docs.python.org/3/tutorial/introduction.html#strings
- [`substring()`](https://www.ibm.com/docs/en/db2/11.5?topic=functions-substring) in IBM Db2
- [`substring()`](https://mariadb.com/kb/en/substring/) in MariaDB
- [`substring()`](https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_substring) In MySQL
- [`substring()`](https://docs.oracle.com/en/database/other-databases/nosql-database/22.1/sqlreferencefornosql/substring_function.html) in Oracle Database
- [`substring()`](https://www.postgresql.org/docs/current/functions-string.html)  in PostgreSQL
- [`substring()`](https://www.sqlite.org/lang_corefunc.html#substr) in SQLite
- [`substring()`](https://learn.microsoft.com/en-us/sql/t-sql/functions/substring-transact-sql?view=sql-server-ver16) in Microsoft SQL Server

**Comment by max-sixty** [comment]
Yes I agree this would be good. 

I'm not sure what the best syntax for this is. If we want to use something like an indexer with `[2:5]`, then we have to know the type — whether we're indexing a string or an array, for example.

So we could start this with a `substring` function, and then later promote that to some indexing syntax when we're more confident on how that should work with types.

**Reply by vanillajonathan** [reply]
I think it would be preferable to use the same bracket notation as Python as many are familiar with it. Follow the [principle of least astonishment](https://en.wikipedia.org/wiki/Principle_of_least_astonishment) (POLA).

**Reply by eitsupi** [reply]
I think `[]` is confusing because it is already used as a list.
https://github.com/PRQL/prql/blob/a82429db6ea1b53c5638ce46f769cb3b0265c148/book/src/syntax.md?plain=1#L18

**Reply by vanillajonathan** [reply]
PHP and JavaScript also uses `[]` notation for lists but also use it for arrays. (PHP: [reference](https://www.php.net/manual/en/language.types.array.php), JS: [reference](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array))

**Reply by eitsupi** [reply]
Currently `from a | select [b]`  cannot be written as `from a | select[b]`.
Doesn't it confuse people that the two uses have different about space?

**Reply by vanillajonathan** [reply]
I don't know. Maybe it does, because in PHP and JavaScript there is a equal sign when assigning an array, and no word before when passing an array into a function.
```javascript
let fruits = ["apple", "banana", "pear", "orange"];
let fruit = fruits[1];
eat_fruits(["apple", "banana", "pear", "orange"]);
```

**Reply by vanillajonathan** [reply]
#3913 adds support for the `substring` function.
